### PR TITLE
Replace md5 hashing with sha512

### PR DESCRIPTION
**Pull request by ioparaskev**
Use sha512 instead of md5 for any hashing operation. MD5 is completely broken and should not be used for anything related to password. Even using MD5 to share password hashes should be treated as if giving the hashes in plaintext.

### CODE REVIEW

**Comment by jirikuncar** [comment]
@ioparaskev I agree we should not use `md5` for password related operations. In case of tokens, the `md5` is used to hash password hashes so it should not be an immediate thread. Can you update the PR in a way in which it can handle tokens generated with both hashing functions similarly to deprecated password hashing functions (see https://github.com/mattupstate/flask-security/blob/develop/flask_security/core.py#L249-L257)?

Maybe something along these lines:
```python
from passlib.context import CryptContext
ctx = CryptContext(schemes=['hex_sha256', 'hex_md5'], deprecated=['hex_md5'])
ctx.verify(user.password, password_hash)
```

**Comment by ioparaskev** [comment]
@jirikuncar It's not an immediate risk, although using md5 for operations where a password hash is involved is like sending the password hash. MD5 hashing a password hash does not add any security at all. So since we don't want the password hashes to be publicly available, we should also not send them hashed with md5.

I was also thinkg instead of the hexdigest, to use the salted version to avoid precomputed attacks.
So I was thinking something like the following:
````python
from passlib.context import CryptContext
ctx = CryptContext(schemes=['sha512_crypt', 'hex_md5'], deprecated=['hex_md5'])
#etc...
````

**Comment by jirikuncar** [comment]
> I was also thinkg instead of the hexdigest, to use the salted version to avoid precomputed attacks.

@ioparaskev Sounds good 👍

File: flask_security/core.py [hunk 1]
```diff
@@ -198,7 +196,7 @@ def _token_loader(token):
         data = _security.remember_token_serializer.loads(
             token, max_age=_security.token_max_age)
         user = _security.datastore.find_user(id=data[0])
-        if user and safe_str_cmp(md5(user.password), data[1]):
+        if user and (verify_hash(data[1], user.password)):
```
Comment (jirikuncar) [inline flask_security/core.py]: You can remove the bracket around the `verify_hash` call.

File: flask_security/utils.py [hunk 2]
```diff
@@ -163,8 +164,17 @@ def encode_string(string):
     return string
 
 
-def md5(data):
-    return hashlib.md5(encode_string(data)).hexdigest()
+def hash_data(data):
+    ctx = CryptContext(schemes=['sha512_crypt', 'hex_md5'],
+                       deprecated=['hex_md5'])
+    return ctx.hash(encode_string(data))
+    # return hashlib.sha512(encode_string(data)).hexdigest()
```
Comment (jirikuncar) [inline flask_security/utils.py]: Please remove the commented line. Thank you.

**Comment by ioparaskev** [comment]
do you think anything else is needed?

**Comment by ioparaskev** [comment]
@jirikuncar do you think anything else is needed

**Comment by jirikuncar** [comment]
@ioparaskev it has been integrated to Flask-Security-Fork as https://github.com/inveniosoftware/flask-security-fork/pull/30. There was no answer from current maintainers in https://github.com/mattupstate/flask-security/issues/559 so I can't merge it here 😞

**Pull request merged**